Make Redpanda to reject null keys in a compacted topic

rystsov opened the pull request:
Kafka doesn't support null keys in a compacted topics. Updating Redpanda to reject it too.

fixes: https://github.com/vectorizedio/support/issues/117

mmaslankaprv commented:
I would split first commit into: 
- adding new fields to `model::record_batch`
- adding new APIs to segment 
- adding decompression to kafka layer

mmaslankaprv commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -279,37 +279,63 @@ static partition_produce_stages produce_topic_partition(
+                      std::move(reader),
+                      acks,
+                      num_records);
+                    return stages.dispatched
+                      .then_wrapped(
+                        [source_shard, dispatch = std::move(dispatch)](
+                          ss::future<> f) mutable {
+                            if (f.failed()) {
+                                (void)ss::smp::submit_to(
+                                  source_shard,
+                                  [dispatch = std::move(dispatch),
+                                   e = f.get_exception()]() mutable {
+                                      dispatch->set_exception(e);
+                                      dispatch.reset();
+                                  });
+                                return;
+                            }
+                            (void)ss::smp::submit_to(
+                              source_shard,
+                              [dispatch = std::move(dispatch)]() mutable {
+                                  dispatch->set_value();
+                                  dispatch.reset();
+                              });
+                        })
+                      .then([f = std::move(stages.produced)]() mutable {
+                          return std::move(f);
+                      });
+                });
```
This part did not change, i think this would befit if we would split formatting and code changes.

mmaslankaprv commented on src/v/model/record.h:
```diff
@@ -647,16 +670,32 @@ class record_batch {
+      record_batch_header header,
+      iobuf&& records,
+      bool compressed,
+      ss::lw_shared_ptr<std::vector<key_offset>> key_offsets)
```
it doesn't have to be shared pointer, `model::record_batch` is not copyable

rystsov replied:
yes but we still copy it by calling `.share()` inside `src/v/raft/consensus_utils.cc` so if we make it a vector we'd need to deep copy instead of just copying a pointer

mmaslankaprv replied:
I would say that it would be more safe to make a deep copy when it comes to cross core accesses. Also this is the thing to consider if we need to wrap `key_offsets` with `foreign_pointer`

mmaslankaprv commented on src/v/model/record.h:
```diff
@@ -647,16 +670,32 @@ class record_batch {
+        _key_offsets = ss::make_lw_shared<std::vector<key_offset>>(
+          std::move(key_offsets));
+    }
+    const ss::lw_shared_ptr<std::vector<key_offset>> key_offsets() const {
```
this method can return `const std::vector<key_offset>&`

rystsov replied:
do you mean `const std::vector<key_offset>& key_offsets() const { return *_key_offsets; }`?

senior7515 replied:
+1 (if we know they are always there)

mmaslankaprv commented on src/v/model/record.h:
```diff
@@ -350,6 +350,17 @@ class record_batch_attributes final {
     std::bitset<16> _attributes;
 };
 
+struct key_offset {
```
maybe we can change this name to in the `record_...` convention, we have `record`, `record_batch`, maybe `record_key_offset` ?

mmaslankaprv commented:
would be great to add some test either in C++or ducktape or both actually.

mmaslankaprv commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -279,37 +279,63 @@ static partition_produce_stages produce_topic_partition(
+                    f = storage::internal::decompress_batch(batch).then(
+                      [&batch](model::record_batch&& b) {
+                          return model::for_each_record(
+                            b, [&batch](const model::record& r) {
```
I think capturing `&batch` may lead to accessing it after it was freed.

rystsov replied:
It is captured by ref inside f, but after `f` is executed it is captured with move `f.then([.. batch = std::move(batch) ..])` and I thought the latter kept batch alive while `f` is executing. Do think this logic is flawed?

mmaslankaprv replied:
I think after the batch is moved reference is invalidated

senior7515 commented:
Can you comment on the additional mem pressure. Why not keep a list of xxhash32 instead of materialized bytes in the record-batch-keys since you can always drop to byte comparison

senior7515 commented on src/v/model/record.h:
```diff
@@ -350,6 +350,17 @@ class record_batch_attributes final {
+
+    int32_t offset_delta;
+    int32_t key_size;
+    bytes key;
```
this seems very expensive. This struct is 48 bytes, per key. so a batch of 1K records is now over a full page. 

I think you can store this on a simpler struct. 

```
struct record_batch_key_offset {

    int32_t offset_delta; 
    int32_t key_size;  ( alex is not sure why we need the key_size)
    int64_t key_hash;
};

```

Is there a world where we use the key_hash - say xxhash64 or some other good non crypto hash that degrades into a *secondary batch decompression* if and only if there is a collision?

rystsov replied:
We copy a batch [here](https://github.com/vectorizedio/redpanda/pull/2435/commits/517a4ad490870491981e11728059c43260d1195d#diff-057947c463347b74bf5eef62d2b15f6efcc2ff244c53ac3f93892a4a9ff0546fR190) but avoid copying it [there](https://github.com/vectorizedio/redpanda/blob/076105985c9ddf2d7056350907da5a30cfaa34c9/src/v/model/record_batch_reader.h#L115) by switching from make_foreign_memory_record_batch_reader to make_memory_record_batch_reader

We used to have a compressed/decompressed batch at the same time [there](https://github.com/vectorizedio/redpanda/pull/2435/commits/99eb85f3f211d6651d18beee42718757f8862fcc#diff-7a75bec80b6c419b49048c0bcc83a6d08bdf9f02288789393a4c2794dfb3ed15R366) and now we decompress and cache it [here](https://github.com/vectorizedio/redpanda/pull/2435/commits/517a4ad490870491981e11728059c43260d1195d#diff-057947c463347b74bf5eef62d2b15f6efcc2ff244c53ac3f93892a4a9ff0546fR187) so memory pressure looks the same.

senior7515 commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -279,37 +279,104 @@ static partition_produce_stages produce_topic_partition(
-                  });
+                      });
+                }
+                return f.then([partition = std::move(partition),
```
should we make some of these `.then()` separate free functions. Seems hard to read at a glance due to future nesting.

rystsov commented:
> I would split first commit into:
> 
>     * adding new fields to `model::record_batch`
> 
>     * adding new APIs to segment
> 
>     * adding decompression to kafka layer

done

rystsov commented:
> would be great to add some test either in C++or ducktape or both actually.

good idea, working on it

rystsov commented:
> would be great to add some test either in C++or ducktape or both actually.

@mmaslankaprv I've added the tests!

senior7515 commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -190,6 +179,31 @@ static partition_produce_stages partition_append(
 }
 
+struct record_batch_pair {
+    model::record_batch batch;
```
Does the cover letter address this design. It seemed to have swapped from caching keys to full copy. 

Means at least 2x. Which is ok as long as short lived. One thing to take into account is if we have to put more back pressure on the Kafka layer give That a compacted topic is at least 2x mem ? 

In the semaphore reservation of the RPC it seems like some accounting is missing.

mmaslankaprv commented on src/v/storage/disk_log_appender.cc:
```diff
@@ -106,7 +106,7 @@ disk_log_appender::operator()(model::record_batch& batch) {
 }
 
 ss::future<ss::stop_iteration>
-disk_log_appender::append_batch_to_segment(const model::record_batch& batch) {
```
maybe if `model::record_batch` interface would return const reference to decompressed record batch (it can be stored as a `std::optional` it would not be necessary to change this method signature. 
`model::record_batch` interface may look like:
```c++
class record_batch {
public:

bool has_decompressed() const {
  return decompressed.has_value();
}

const model::record_batch& decompressed() const {
  return *decompressed;
}

model::record_batch release_decompressed() && {
   return std::move(*decompressed);
}


private:
std::optional<model::record_batch> decompressed;
}

rystsov replied:
As I understand if I make it an option then I won't be able to share it [here](https://github.com/vectorizedio/redpanda/pull/2435/commits/0d24d7c226cb5356afc48cd9f625836a1ae92896#diff-b54120497a6a46905cf09c97f78217463ad1cf23dfc322cec34b5c1b7b92cd19R611) and would need to copy it and thus increase memory pressure.

mmaslankaprv replied:
can we do it like: 
```
    record_batch share() {
        return record_batch(
          _header,
          _records.share(0, _records.size_bytes()),
          _compressed,
          _decompressed.has_value() ? std::optional<model::record_batch>(_decompressed->share()) : std::nullopt);
    }
```

mmaslankaprv replied:
good point, replacing `std::optional` with `std::unqiue_ptr<model::record_batch>` will help

LenaAn commented:
Does this PR still needs a review or is it obsolete?

mmaslankaprv reviewed (commented):
I left two minor comments. Would it be possible to add happy path test with compression to the test suite ?

rystsov commented:
> I left two minor comments. Would it be possible to add happy path test with compression to the test suite ?

@mmaslankaprv  I've fixed all the nits and added a couple of positive tests to the suite. Unfortunately I've messed with `git merge` and had to rebase

mmaslankaprv commented:
this looks great, the only thing left before merging is to fix `compacted_topic_recovery` test which failed consistently in all build scenarios.

dotnwat commented on src/v/model/record.h:
```diff
@@ -550,8 +550,11 @@ class record_batch {
-        return record_batch(_header.copy(), _records.copy(), _compressed);
+        auto b = record_batch(_header.copy(), _records.copy(), _compressed);
+        if (_decompressed) {
+            b.decompressed(_decompressed->copy());
```
```
vassert(!_decompressed->compressed(), "");
b.decompressed(_decompressed->copy());
```

would that make sense to add as an assertion?

rystsov replied:
good idea, done!

dotnwat commented on src/v/cluster/partition.h:
```diff
@@ -137,6 +137,8 @@ class partition {
         return _raft->linearizable_barrier();
```
```
compaction: decompress batch in kafka layer
```
can you describe the strategy for how data is moving across cores? which core incurs the cost of decompression, etc?

rystsov replied:
We dispatch a produce request to the right core pretty early in the processing pipeline, see `produce_topic_partition` function of `produce.cc`. It allows us to move decompression from the storage layer (segment.cc) to the kafka layer without crossing the core boundary so it still happens on the same core as it used before.

dotnwat commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -190,6 +179,63 @@ static partition_produce_stages partition_append(
+    std::optional<produce_response::partition> error_opt;
+
+    std::vector<int32_t> null_keys;
+    if (batch.decompressed()) {
+        batch.decompressed()->for_each_record(
+          [&null_keys](const model::record& r) {
+              if (r.key_size() < 0) {
+                  null_keys.push_back(r.offset_delta());
+              }
+          });
+    } else if (!batch.compressed()) {
```
can we work on the record_batch interface to make this a little more intuitive to read? For example, `if (batched.decompressed())` reads as if it is equivalent to the else condition `if (!batch.compressed())`

dotnwat commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -190,6 +179,63 @@ static partition_produce_stages partition_append(
+        return ss::do_with(std::move(batch), [](model::record_batch& batch) {
+            return storage::internal::decompress_batch(batch).then(
+              [&batch](model::record_batch&& b) {
+                  batch.decompressed(std::move(b));
```
from what I can tell once we set the decompressed representation here it follows around the batch through copies, shares, etc... how does this affect memory usage in the batch cache? it looks like if we hold on to a batch with decompressed data in the cache we'll have 2 copies of the data?

dotnwat reviewed (commented):
overall strategy seems great. i think there seems to be some full batch copy/copies that could be avoided and I can't quite see how this interacts with the batch cache such as are we going to cache compressed and uncompressed data there? also looks like we could avoid storing two batch headers in the record batch for the decompressed version.
